system/include/compat/avxintrin.h

```diff
@@ -11,14 +11,505 @@
+  __m256d ret;
+  ret.v0 = _mm_permutevar_pd(__a.v0, __c.v0);
+  ret.v1 = _mm_permutevar_pd(__a.v1, __c.v1);
+  return ret;
+}
+
+static __inline __m128 __DEFAULT_FN_ATTRS _mm_permutevar_ps(__m128 __a,
+                                                            __m128i __c) {
+  return (__m128)wasm_f32x4_make(
+    ((__f32x4)__a)[wasm_i32x4_extract_lane(__c, 0) & 3],
+    ((__f32x4)__a)[wasm_i32x4_extract_lane(__c, 1) & 3],
+    ((__f32x4)__a)[wasm_i32x4_extract_lane(__c, 2) & 3],
+    ((__f32x4)__a)[wasm_i32x4_extract_lane(__c, 3) & 3]);
+}
+
+static __inline __m256 __DEFAULT_FN_ATTRS _mm256_permutevar_ps(__m256 __a,
+                                                               __m256i __c) {
+  __m256 ret;
+  ret.v0 = _mm_permutevar_ps(__a.v0, __c.v0);
+  ret.v1 = _mm_permutevar_ps(__a.v1, __c.v1);
+  return ret;
+}
+
+#define _mm_permute_pd(__a, __imm)                                             \
+  __extension__({                                                              \
+    (__m128d) wasm_i64x2_shuffle(                                              \
+      (__m128d)(__a), (__m128d)(__a), ((__imm) & 1), (((__imm) >> 1) & 1));    \
+  })
+
+#define _mm256_permute_pd(__a, __imm)                                          \
+  __extension__({                                                              \
+    _mm256_set_m128d(_mm_permute_pd((__a).v1, (__imm) >> 2),                   \
+                     _mm_permute_pd((__a).v0, (__imm)));                       \
+  })
+
+#define _mm_permute_ps(__a, __imm)                                             \
+  __extension__({                                                              \
+    (__m128) wasm_i32x4_shuffle((__m128)(__a),                                 \
+                                (__m128)(__a),                                 \
+                                ((__imm) & 3),                                 \
+                                (((__imm) >> 2) & 3),                          \
+                                (((__imm) >> 4) & 3),                          \
+                                (((__imm) >> 6) & 3));                         \
+  })
+
+#define _mm256_permute_ps(__a, __imm)                                          \
+  __extension__({                                                              \
+    _mm256_set_m128(_mm_permute_ps((__a).v1, (__imm)),                         \
+                    _mm_permute_ps((__a).v0, (__imm)));                        \
+  })
+
+static inline __m128d select4d(__m256d __a, __m256d __b, const int imm8) {
+  const int index = imm8 & 0x3;
+  __m128d tmp;
+  switch (index) {
+    case 0:
+      tmp = __a.v0;
+      break;
+    case 1:
+      tmp = __a.v1;
+      break;
+    case 2:
+      tmp = __b.v0;
+      break;
+    case 3:
+      tmp = __b.v1;
+      break;
+  }
+  if (imm8 & 0x8) {
+    tmp = (__m128d)wasm_i64x2_const_splat(0);
+  }
```

Comment:
Even though Intel intrinsics guide does not well-define what should happen if `imm8` has value of 4,5,6 or 7 (making these functions return an uninitialized `tmp`), I think it would be good to include that in testing so that in our unit test suite, the practical behavior will align with what a native compiled AVX intrinsics using program would do under these inputs.

Also, while having an implementation that diligently follows the reference is nice, I would recommend fusing the `switch` and `if` statements into a single check, i.e. try something like

```c
static __inline__ __m128d __attribute__((__always_inline__, __nodebug__))
__avx_select4d(__m256d __a, __m256d __b, const int imm8)
{
  switch(index & 0x11) {
  case 0: return __a.v0;
  case 1: return __a.v1;
  case 2: return __b.v0;
  case 3: return __b.v1;
  default: return (__m128d)wasm_i64x2_const_splat(0);
}
```
and similar for the other `select4` variants. And tune the cases depending on what native AVX code does if an `imm8` of 4,5,6 or 7 is passed.

Reply:
Bit 3 and bit 0/1 have different priorities. if bit 3 is set, select4 return zero regardless of what the other bits are, so we cannot fusing the switch and if statements. I will pull if before switch, which may make this more clear.

Reply:
> Bit 3 and bit 0/1 have different priorities. if bit 3 is set, select4 return zero regardless of what the other bits are, so we cannot fusing the switch and if statements.

I thought the above switch-case takes that into account? If bit 3 is set, then `index & 0x11` is >= 8, so it will always go down the `default:` path?

Reply:
Gotcha, thanks! New commit is uploaded, address all the review comments.